lib/commands/audit.js

```diff
@@ -1,8 +1,384 @@
+    const nodes = this.tree.inventory.values()
+    this.getEdges(nodes, 'edgesOut')
+    const edges = Array.from(this.edges)
+    if (edges.length === 0) {
+      throw new Error('No dependencies found')
+    }
+
+    // Prefetch and cache public keys from used registries
+    const registries = this.findAllRegistryUrls(edges, this.npm.flatOptions)
+    for (const registry of registries) {
+      const keys = await this.getKeys({ registry })
+      if (keys) {
+        this.keys.set(registry, keys)
+      }
+    }
+
+    await Promise.all(edges.map((edge) => this.getVerifiedInfo(edge)))
+
+    // Sort alphabetically
+    const invalid = Array.from(this.invalid).sort((a, b) => localeCompare(a.name, b.name))
+    const missing = Array.from(this.missing).sort((a, b) => localeCompare(a.name, b.name))
+
+    const verified = invalid.length === 0 && missing.length === 0
+
+    if (!verified) {
+      this.exitCode = 1
+    }
+
+    const end = process.hrtime.bigint()
+    const elapsed = end - start
+
+    if (this.npm.config.get('json')) {
+      this.appendOutput(this.makeJSON({ invalid, missing }))
+    } else {
+      const timing = `audited ${this.audited} packages in ${Math.floor(Number(elapsed) / 1e9)}s`
+      const verifiedPrefix = verified ? 'verified registry signatures, ' : ''
+      this.appendOutput(`${verifiedPrefix}${timing}\n`)
+
+      if (this.verified && !verified) {
+        this.appendOutput(
+          `${this.verified} packages have ${chalk.bold('verified')} registry signatures\n`
+        )
+      }
+
+      if (missing.length) {
+        const msg = missing.length === 1 ?
+          `package has a ${chalk.bold(chalk.magenta('missing'))} registry signature` :
+          `packages have ${chalk.bold(chalk.magenta('missing'))} registry signatures`
+        this.appendOutput(
+          `${missing.length} ${msg} but the registry is ` +
+          `providing signing keys${this.npm.config.get('missing') ? ':\n' : ''}`
+        )
+        // TODO: This might not be the right option for this
+        if (this.npm.config.get('missing')) {
+          this.appendOutput(this.humanOutput(missing))
+        } else {
+          this.appendOutput(`  run \`npm audit signatures --missing\` for details`)
+        }
+      }
+
+      if (invalid.length) {
+        const msg = invalid.length === 1 ?
+          `package has an ${chalk.bold(chalk.red('invalid'))} registry signature` :
+          `packages have ${chalk.bold(chalk.red('invalid'))} registry signatures`
+        this.appendOutput(
+          `${missing.length ? '\n' : ''}${invalid.length} ${msg}:\n`
+        )
+        this.appendOutput(this.humanOutput(invalid))
+        const plural = invalid.length === 1 ? '' : 's'
+        this.appendOutput(
+          `\nSomeone might have tampered with the package${plural} ` +
+          `since it was published on the registry (monster-in-the-middle attack)!\n`
+        )
+      }
+    }
+  }
+
+  findAllRegistryUrls (edges, opts) {
```

Comment:
@wraithgar this func doesn't seem ideal to me but unsure how to improve. Is there ever a case where transitive deps have a different registry to top level ones? Could we just get the registry for top-level deps?